File: lambda/get_presigned_url.py
import json
import boto3
import os
import uuid

s3_client = boto3.client("s3", 
                         region_name=os.environ.get("REGION", "eu-central-1"),
                         endpoint_url=f"https://s3.{os.environ.get('REGION','eu-central-1')}.amazonaws.com")

# Read bucket name from environment variable
BUCKET_NAME = os.environ.get("UPLOAD_BUCKET", "s3-bucket-name")

corsHeaders = {
    "Content-Type": "application/json",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type",
    "Access-Control-Allow-Methods": "GET,OPTIONS,PUT",
}

def lambda_handler(event, context):
    """
    Lambda handler that generates a presigned S3 URL for upload.
    """

    headers = event.get("headers", {})
    custom_header = headers.get("x-custom-auth")

    if custom_header != os.environ.get("CUSTOM_AUTH_SECRET"):
        return {
            "headers": corsHeaders,
            "statusCode": 403,
            "body": json.dumps({"error": "Forbidden: Invalid or missing custom auth header"})
        }
    
    try:
        # Generate a random unique filename (you can also prefix with user ID or folder)
        file_key = str(uuid.uuid4())

        presigned_url = s3_client.generate_presigned_url(
            "put_object",
            Params={
                "Bucket": BUCKET_NAME,
                "Key": file_key
            },
            ExpiresIn=3600  # URL valid for 1 hour
        )

        return {
            "statusCode": 200,
            "headers": corsHeaders,
            "body": json.dumps({
                "upload_url": presigned_url,
                "file_key": file_key
            })
        }

    except Exception as e:
        return {
            "headers": corsHeaders,
            "statusCode": 500,
            "body": json.dumps({"error": str(e)})
        }

File: modules/image_uploader/lambda_get_presigned_url/getPresignedUrl.js
const AWS = require("aws-sdk");
const crypto = require("crypto");

const REGION = process.env.REGION || "eu-central-1";
const BUCKET_NAME = process.env.UPLOAD_BUCKET || "s3-bucket-name";
const UPLOAD_FOLDER = process.env.UPLOAD_FOLDER || "uploads/";
const EXPIRATION_TIME_S = parseInt(process.env.EXPIRATION_TIME_S || "300");
const API_GW_AUTH_SECRET = process.env.API_GW_AUTH_SECRET;
const PARTITION_KEY = process.env.PARTITION_KEY || "user_id";
const SORT_KEY = process.env.SORT_KEY || "file_key";

const s3 = new AWS.S3({
    region: REGION,
    signatureVersion: "v4",
    useAccelerateEndpoint: process.env.USE_S3_ACCEL === "true"
});

const corsHeaders = {
    "Content-Type": "application/json",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type,x-api-gateway-img-upload-auth",
    "Access-Control-Allow-Methods": "GET,OPTIONS,PUT",
};

exports.handler = async (event) => {
    const headers = event.headers || {};
    const customHeader = headers["x-api-gateway-img-upload-auth"];

    if (customHeader !== API_GW_AUTH_SECRET) {
        return {
            statusCode: 403,
            headers: corsHeaders,
            body: JSON.stringify({ error: "Forbidden: Invalid or missing custom auth header" }),
        };
    }

    const query = event.queryStringParameters || {};

    // Client must send tripId, filename, extension, resource (name in plurals)
    const partitionKey = query[PARTITION_KEY];
    const originalFilename = query[SORT_KEY];
    const extension = query.ext;
    const apiResource = query.resource;

    if (!partitionKey || !originalFilename || !extension) {
        return {
            statusCode: 400,
            headers: corsHeaders,
            body: JSON.stringify({ error: "Query params could be missing" })
        };
    }

    try {
        // Generate a random unique filename
        const randomId = crypto.randomBytes(16).toString("base64url"); // URL-safe base64

        // Build key: uploads/<resource>/<partition_key>/<randomId>_<filename>.<extension>
        const fileKey = `${UPLOAD_FOLDER}${apiResource}/${partitionKey}/${randomId}_${originalFilename}${extension ? "." + extension : ""}`;

        // Generate presigned PUT URL
        const presignedUrl = s3.getSignedUrl("putObject", {
            Bucket: BUCKET_NAME,
            Key: fileKey,
            Expires: EXPIRATION_TIME_S
        });

        return {
            statusCode: 200,
            headers: corsHeaders,
            body: JSON.stringify({
                upload_url: presignedUrl,
                file_key: fileKey
            })
        };

    } catch (error) {
        return {
            statusCode: 500,
            headers: corsHeaders,
            body: JSON.stringify({ error: error.message })
        };
    }
};
